### PR TITLE
rust: Store absolute paths in rust-project.json

### PR DESCRIPTION
**Pull request by sdroege**
As meson requires source_dir!=build_dir and stores the rust-project.json inside the build directory, while software like rust-analyzer expects it at the root of the source directory, manual steps are needed for making them work together.

One option, as described in the documentation, is per project configuration. Another option, that works correctly with compile-commands.json and clangd, is to store a symlink to the file in the build directory at the root of the source directory.

As currently rust-project.json stores paths relative to the location of the file itself and rust-analyzer does not resolve symlinks, this does not work.

To solve this, store absolute paths in rust-project.json as is already done in compile_commands.json for the directory.

----

rust-analyzer PR to resolve symlinks: https://github.com/rust-lang/rust-analyzer/pull/14168

CC @dcbaker

### CODE REVIEW

**Comment by sdroege** [comment]
See also https://github.com/rust-lang/rust-analyzer/pull/14168

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -462,7 +462,7 @@ class RustCrate:
     def to_json(self) -> T.Dict[str, object]:
         ret: T.Dict[str, object] = {
             "display_name": self.display_name,
-            "root_module": self.root_module,
+            "root_module": str(Path(self.root_module).resolve()),
```
[the remark this answers is not in the document]
Comment (sdroege) [reply mesonbuild/backend/ninjabackend.py:465]: I thought pathlib was generally recommended over the `os.path` API but I can change that :)

**Comment by dcbaker** [comment]
the changes themselves look good to me. Thanks for looking at this, and especially taking the issue upstream :)

I (unfortunately, because I find `pathlib` to much nicer than `os.path`), also find it to have an unfortunate set of bugs and problematic behaviors that make it not actually as useful as it should be.

With the pathlib -> os.path changes this is r-b from me

**Comment by nirbheek** [comment]
> I (unfortunately, because I find `pathlib` to much nicer than `os.path`), also find it to have an unfortunate set of bugs and problematic behaviors that make it not actually as useful as it should be.

Completely agree. I was all aboard the `pathlib` train till I started to actually use it, and now I'm fully off it.

**Comment by sdroege** [comment]
Updated but don't merge this yet. The problem right now is that if you run meson from inside the build directory it writes the correct path, if you run it from the source directory it won't (and will put a path relative to the source directory).

**Comment by sdroege** [comment]
I think this should be fixed directly in `generate_rust_target` instead.

**Comment by sdroege** [comment]
This should work fine. For the proc macro dylib path this should be unnecessary as it seems to generate an absolute path already.

**Comment by sdroege** [comment]
Also fixed the proc-macro handling there in another commit. That was completely broken but now the `test cases/rust/18 proc-macro/` testcase works correctly.

**Review by nirbheek — APPROVED** [review]

**Comment by sdroege** [comment]
`test cases/rust/11 generated main` needs fixing

**Comment by sdroege** [comment]
Works here now but for some of the test cases rust-analyzer doesn't find dependencies (before and after my changes).

**Pull request merged**